python/ray/data/_internal/stats.py

```diff
@@ -339,61 +346,131 @@ def _check_cluster_stats_actor():
-    global _stats_actor
-    _check_cluster_stats_actor()
+    Currently there are two codepaths that report stats to _StatsActor.
+        1. Execution:
+            Updates are called from the driver process in _scheduling_loop_step.
+            An update is passed through to _StatsActor via remote call if the last
+            update was more than `STATS_ACTOR_UPDATE_INTERVAL_SECONDS` seconds ago.
+        2. Iteration:
+            Updates are called from wherever iteration takes place in iter_batches.
+            Theoretically it is possible (not common) for multiple iterators to make
+            updates, so we batch the latest update for each iterator and make a remote
+            call to _StatsActor every `STATS_ACTOR_UPDATE_INTERVAL_SECONDS` seconds.
+            The remote call is made in a background thread that sleeps for 5 seconds
+            after calling. This thread will automatically close after each iterator
+            clears their metrics and after `STATS_ACTOR_UPDATE_THREAD_INACTIVITY_LIMIT`
+            iterations of inactivity.
 
-    _stats_actor.update_iter_metrics.remote(stats, tags_list)
+    """
 
+    def __init__(self):
+        # Time of last update from execution codepath
+        self._last_update: float = 0
 
-def clear_stats_actor_metrics(tags_list: List[Dict[str, str]]):
-    global _stats_actor
-    _check_cluster_stats_actor()
+        # Last stats snapshots for each dataset
+        self._last_iter_stats: Dict[str, Tuple[Dict[str, str], "DatasetStats"]] = {}
+        # Background thread to make remote calls to _StatsActor
+        self._iter_update_thread: Optional[threading.Thread] = None
+        # Keeps track of number of inactive iterations for the background thread.
+        self._iter_stats_inactivity: int = 0
 
-    _stats_actor.clear_metrics.remote(tags_list)
+    # Execution methods
 
+    def update_stats_actor_metrics(
+        self,
+        op_metrics: List[OpRuntimeMetrics],
+        tags_list: List[Dict[str, str]],
+        state: Dict[str, Any],
+        force_update: bool = False,
+    ):
+        if (
+            force_update
+            or time.time() - self._last_update >= STATS_ACTOR_UPDATE_INTERVAL_SECONDS
+        ):
+            global _stats_actor
+            _check_cluster_stats_actor()
+            _stats_actor.update_metrics.remote(
+                [metric.as_dict() for metric in op_metrics], tags_list, state
+            )
+            self._last_update = time.time()
+
+    def clear_stats_actor_metrics(self, tags_list: List[Dict[str, str]]):
+        global _stats_actor
+        _check_cluster_stats_actor()
 
-def clear_stats_actor_iter_metrics(tags: Dict[str, str]):
-    global _stats_actor
-    _check_cluster_stats_actor()
+        _stats_actor.clear_metrics.remote(tags_list)
 
-    _stats_actor.clear_iter_metrics.remote(tags)
+    # Iteration methods
 
+    def update_stats_actor_iter_metrics(
+        self, stats: "DatasetStats", tags: Dict[str, str]
+    ):
```

Comment:
should we also force-update when the iterator finishes?

Reply:
The point of force-update for the other codepath is to update the `_StatsActor.datasets`, not to update the metrics, since we clear the metrics right after anyways, its unlikely those last metrics will even be emitted since prometheus only samples every couple seconds

Reply:
This is a separate issue that would require support for actually stopping the emitting of a metric for a certain tag (right now we just set the metric to 0). A simple way to make sure that the last metrics get emitted is to also just wait 10s before clearing.